### PR TITLE
fix clusterizer to handle negative charge

### PR DESCRIPTION
**Pull request by dkotlins**
This is a one line code addition to the pixel clusterizer to take care of very low pixel charges.
It has to be accompanied by  the lowering of all pixel clusterizer thresholds. Especially the pixel 
threshold (called channelThreshold) has to be below 100 in order to recover low charge pixels.

### CODE REVIEW

**Comment by boudoul** [comment]
Dear @perrotta and @slava77 , please note that this fix is important to get towards the next re-reco of 2017 data 
@venturia @veszpv  FYI

File: RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc [hunk 1]
```diff
@@ -266,12 +266,17 @@ void PixelThresholdClusterizer::copy_to_buffer( DigiIterator begin, DigiIterator
...
-    int adc = electron[i++];
+    int adc = electron[i++]; // this is in electrons 
+
 #ifdef PIXELREGRESSION
     int adcOld = calibrate(di->adc(),col,row);
     //assert(adc==adcOld);
     if (adc!=adcOld) std::cout << "VI " << eqD  <<' '<< ic  <<' '<< end-begin <<' '<< i <<' '<< di->adc() <<' ' << adc <<' '<< adcOld << std::endl; else ++eqD;
 #endif
+
+    //if(adc<0) cout<<" negative amplitude "<<adc<<" "<<row<<" "<<col<<" "<<calibrate(di->adc(),col,row)<<" "<<detid_<<endl;
```
Comment (perrotta) [inline RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc]: Please remove unneded commented out code

File: RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc [hunk 2]
```diff
@@ -324,7 +329,9 @@ int PixelThresholdClusterizer::calibrate(int adc, int col, int row)
 	  //const float pedestal = -28.2 * gain; // -79.
 	  
 	  float DBgain     = theSiPixelGainCalibrationService_->getGain(detid_, col, row);
-	  float DBpedestal = theSiPixelGainCalibrationService_->getPedestal(detid_, col, row) * DBgain;
+	  float pedestal   = theSiPixelGainCalibrationService_->getPedestal(detid_, col, row);
+	  float DBpedestal = pedestal * DBgain;
+	  //float DBpedestal = theSiPixelGainCalibrationService_->getPedestal(detid_, col, row) * DBgain;
```
Comment (perrotta) [inline RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc]: Please remove unneded commented out code

File: RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc [hunk 3]
```diff
@@ -351,6 +358,26 @@ int PixelThresholdClusterizer::calibrate(int adc, int col, int row)
 	  } else {
 	    electrons = int( vcal * theConversionFactor + theOffset); 
 	  }
+
+
+	  // if(adc<0 || adc>255) cout<<" adc wrong range "<<adc<<endl;
```
Comment (perrotta) [inline RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc]: Please remove unneded commented out code, or add a comment which explains why it is needed here

File: RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc [hunk 4]
```diff
@@ -351,6 +358,26 @@ int PixelThresholdClusterizer::calibrate(int adc, int col, int row)
...
+	  // 	(layer_==4  && (pedestal<-90. || pedestal>110.) ) ) 
+
+	  //     cout<<"PED outside range: layer "<<layer_<<" "<<pedestal<<" "<<DBgain<<" "<<DBpedestal
+	  // 	  <<" "<<adc<<" "<<vcal<<" "<<electrons<<" det: "
+	  // 	  <<detid_<<" row: "<<row<<" col: "<<col<<endl;
+
+	  // }
+
+	  if(electrons<0) 
+	    cout<<" layer "<<layer_<<" "<<adc<<" "<<DBgain<<" "<<pedestal<<" "<<DBpedestal<<" "<<vcal
```
Comment (perrotta) [inline RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc]: cout is not thread safe
I suspect anyhow that also this line was originally supposed to get commented out

**Comment by dkotlins** [comment]
Can’t we have any debug code any more? If something goes wrong how do we debug?
Acturally the ifdef PIXEL_REGRESSION is very old and come from Vincenzo I think.
It has been in the code for 5-6 years and nobody has complained until now!
D.

**Comment by perrotta** [comment]
@dkotlins : I didn't comment about "ifdef PIXEL_REGRESSION", only  
about the new lines you added to the code.

Commented out lines should better be avoided. Alternatively, another  
line of comments should explain why they are needed to be kept in the  
code.


dkotlins <notifications@github.com> ha scritto:

> Can’t we have any debug code any more? If something goes wrong how  
> do we debug?
> Acturally the ifdef PIXEL_REGRESSION is very old and come from  
> Vincenzo I think.
> It has been in the code for 5-6 years and nobody has complained until now!
> D.
>

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
@dkotlins 
please add some slides to the PR description which would describe the effect of this PR on physics outputs.
If this affects inputs for tracking in a significant way, some tracking validation plots would be nice as well.

@makortel @VinInn may be able to help with the tracking validation plots.

**Comment by perrotta** [comment]
(Linked to the comment from @slava77 above). Were the large modifications in the thresholds in SiPixelClusterizer_cfi.py already discussed and approved within the Tracking POG? Can you put a link to a presentation where its effect on tracking quantities is shown?

**Comment by perrotta** [comment]
As expected, the jenkins tests show sizeable differences in all reco - related quantities (i.e. tracking anf basically everything higher level). 
Comments and justifications are needed before we can continue with the review.

File: RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc [hunk 5]
```diff
@@ -266,12 +266,16 @@ void PixelThresholdClusterizer::copy_to_buffer( DigiIterator begin, DigiIterator
...
-    int adc = electron[i++];
+    int adc = electron[i++]; // this is in electrons 
+
 #ifdef PIXELREGRESSION
     int adcOld = calibrate(di->adc(),col,row);
     //assert(adc==adcOld);
     if (adc!=adcOld) std::cout << "VI " << eqD  <<' '<< ic  <<' '<< end-begin <<' '<< i <<' '<< di->adc() <<' ' << adc <<' '<< adcOld << std::endl; else ++eqD;
 #endif
+
+    if(adc<100) adc=100; // put all negative pixel charges into the 100 elec bin
```
Comment (fwyzard) [inline RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc]: since the comment is about "negative pixel charges", why `if(adc<100)` rather than `if(adc<0)` ?

Comment (fwyzard) [reply RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc]: wouldn't it make more sense to have the cut value (`adc<100`) and target value (`adc=100`) independent and configurable ?

**Comment by dkotlins** [comment]
I had initially  if(adc<0)adc=100 but corrected it later.
My way of thinking was the following:
say you have a pixel with charge=-1  -> would be assigned to charge=100
another pixel has charge=0 -> would stay with 0 and therefore be lost.
Why would -1 be treated better as 0?
From the detector point of view it would not make much sense.
D.

**Comment by dkotlins** [comment]
Yes, one can do that, but past experience shows that adding new configurable constants 
increases the time to accept the PR.
D.

**Comment by fwyzard** [comment]
Actually, if a module follows the general guidelines for CMSSW (i.e. the use of a `fillDescriptions` method rather than abusing the `existAs` checks), adding a new parameter is straightforward.
I do agree that if one never implements the proper mechanism for handling the module configuration, instead adding one hack over an other, makes the PRs more cumbersome to integrate.

**Comment by dkotlins** [comment]
Yes, there are differences. This is why this code modification was done. It recovers pixels lost dues to negative charge.
It should improve the reconstruction but some distribution will look different  (e.g. cluster charge and size, pixel charge).

**Comment by dkotlins** [comment]
I have gone back with most of the thresholds to their previous values. Only the channel (pixel) threshold is new, 
this has to stay (<100) if we want the fix to work.

**Comment by dkotlins** [comment]
Slides were presented at the Tracker DPG meeting on Tuesday.

**Comment by davidlange6** [comment]
please test

**Comment by perrotta** [comment]
@makortel @VinInn 

This fix does affect reco Tracks and all higher level quantities.

There is the request to integrate it in 940pre2, and have therefore it tested with the needed related payloads (see also https://indico.cern.ch/event/666794/contributions/2729097/attachments/1528713/2391465/tanjaReco_22092017.pdf) .

Did you already manage to make your usual multiTrack validation plots combining this PR with the new payloads, or do you plan to provide some? Could you please comment about the request of integrating this PR in 94X?

Thank you.

**Comment by makortel** [comment]
@perrotta 

> Did you already manage to make your usual multiTrack validation plots combining this PR with the new payloads, or do you plan to provide some?

No (I don't think we've promised such, and anyway to my understanding the effect on MC should be minimal so MTV wouldn't tell much on the effect on data).

> Could you please comment about the request of integrating this PR in 94X?

 Please integrate.

**Comment by perrotta** [comment]
Thank you Matti.

Matti Kortelainen <notifications@github.com> ha scritto:

> @perrotta
>
>> Did you already manage to make your usual multiTrack validation  
>> plots combining this PR with the new payloads, or do you plan to  
>> provide some?
>
> No (I don't think we've promised such, and anyway to my  
> understanding the effect on MC should be minimal so MTV wouldn't  
> tell much on the effect on data).

Indeed, the reference to MTV was incorrect, as the effect is mostly on data.

And yes, you never promised any validation, as far as I know, although  
we repeatedly asked the authors of this PR (even within this github  
thread) to get confirmation from the tracking POG that everything was  
agreed with you, and to provide some possible validation with your help.

>> Could you please comment about the request of integrating this PR in 94X?
>
>  Please integrate.

We are near the latest open 94X pre-release: I take this as a blessing  
of this PR by the tracking POG.

**Comment by perrotta** [comment]
Here are a few validation plots and comparisons from the SinglePh2017B wf 136.788 run by jenkins, to be kept here as a possible low stat reference (only 100 events run by jenkins). The new payloads that must be associated with this PR were not included .

- bTag CVS efficiency of the discriminator cut basically unaffected: 
image

- Increased number of tracking seeds
image
image

- More general tracks, with a lower number of hits per tracks:
image

- Electron gsf track, jets, MET, vertices, muons obviously affected:, but no evidence of altered shapes of the distributions
image
image
image
image
image

- More conversions:
image

**Comment by perrotta** [comment]
+1
- Code wise this fix is quite reasonable, and it has been shown to pinpoint the origin and recover the hit inefficiency in pixels, as shown in https://indico.cern.ch/event/666794/contributions/2729097/attachments/1528713/2391465/tanjaReco_22092017.pdf
- It is endorsed by tracking POG, https://github.com/cms-sw/cmssw/pull/20555#issuecomment-331905010
- Distributions obtained in the jenkins tests look already reasonable (https://github.com/cms-sw/cmssw/pull/20555#issuecomment-331922378), even if those comparisons should be done with the modified pixels conditions. A more detailed validation is certainly needed, though: it is intention of the proponents to do it in 940pre2, where also the needed payloads are expected to be integrated

**Comment by davidlange6** [comment]
+1

**Comment by makortel** [comment]
@perrotta Here is some "post-merge validation" with 2000 events of JetHT run 302131. Full DQM GUI can be found from
http://127.0.0.1:8081/dqm/relval/start?runnr=302131;dataset=/JetHT/CMSSW_9_3_0_pre5_SiPixelGainCalibration_2017_v5_PR20555-v1_2/DQMIO;sampletype=offline_data;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=True;referenceobj1=other%3A%3A/JetHT/CMSSW_9_3_0_pre5_SiPixelGainCalibration_2017_v5-v1_1/DQMIO%3A;referenceobj2=other%3A%3A/JetHT/CMSSW_9_3_0_pre5-Out_of_the_box-v1/DQMIO%3A;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=M;root=Tracking/TrackParameters;focus=;zoom=no;
with `ssh -L8081:mkdev.cern.ch:8081 lxplus.cern.ch` tunnel. Legend
* black: 9_3_0_pre5 + SiPixelGainCalibration_2017_v5 + this PR
* blue: 9_3_0_pre5 + SiPixelGainCalibration_2017_v5
* orange: 9_3_0_pre5

With this PR (wrt. gain v5) we get a tiny bit more quadruplets
image

Resolution improves tiny bit
image

BPix1 hit efficiency (from track hit pattern) improves tiny bit
image

In conclusion the changes look small but are in good direction (as expected). Note that to profit maximally (in resolution) we need also new templates and alignment.
